File: backend/app/schemas/contract.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class ContractCreate(BaseModel):
    booking_id: int


class ContractResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    booking_id: int
    tenant_id: int
    property_id: int
    template_name: str
    agreement_number: str | None = None
    version: int = 1
    template_version: str = "2026.1"
    content_hash: str | None = None
    snapshot: dict[str, Any] | None = None
    generated_at: datetime | None = None
    content: str
    status: str
    signed_at: datetime | None = None
    file_path: str | None = None
    pdf_status: str = "not_generated"
    created_at: datetime
    updated_at: datetime


class SignaturePoint(BaseModel):
    x: float = Field(ge=0, le=1)
    y: float = Field(ge=0, le=1)
    pressure: float = Field(default=0.5, ge=0, le=1)


class ContractSignCreate(BaseModel):
    agreement_version: int = Field(ge=1)
    agreement_content_hash: str = Field(min_length=64, max_length=64)
    tenant_name: str = Field(min_length=1, max_length=200)
    consent_text_version: str = Field(min_length=1, max_length=32)
    idempotency_key: str = Field(min_length=8, max_length=100)
    strokes: list[list[SignaturePoint]] = Field(min_length=1, max_length=100)
    name_confirmed: bool
    electronic_signature_consent: bool


class ContractSignatureResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    agreement_id: str
    agreement_version: int
    agreement_content_hash: str
    tenant_user_id: int
    tenant_name: str
    signed_at: datetime
    property_timezone: str
    consent_text_version: str
    signature_hash: str
    pdf_status: str = "pending"


class TenantContractListItem(BaseModel):
    agreement_id: str
    agreement_number: str
    agreement_version: int
    agreement_content_hash: str
    order_id: str
    booking_id: int
    property_id: int
    tenant_user_id: int
    signed_at: datetime
    lease_start_date: str | None
    lease_end_date: str | None
    lease_months: int | None
    property_timezone: str
    property_name: str
    property_address: str
    property_image_url: str | None
    payment_status: str
    booking_status: str
    reservation_status: str
    agreement_status: str
    category: str
    category_label: str
    status_labels: list[str]
    invalid_reason: str | None = None
    settlement_currency: str | None = None
    settlement_amount_minor: int | None = None
    payment_expires_at: datetime | None = None
    remaining_payment_seconds: int | None = None
    remaining_contract_days: int | None = None
    can_pay: bool = False
    waiting_for_move_in: bool = False
    signed_pdf_available: bool = False


class TenantContractDetail(TenantContractListItem):
    content: str
    snapshot: dict[str, Any]
    signature_url: str


class TenantContractListResponse(BaseModel):
    items: list[TenantContractListItem]
    total: int

File: frontend/src/services/contract.ts
import api from './api'

export interface Contract {
  id: string
  booking_id: number
  tenant_id: number
  property_id: number
  template_name: string
  agreement_number: string | null
  version: number
  template_version: string
  content_hash: string | null
  snapshot: ContractSnapshot | null
  generated_at: string | null
  content: string
  status: string
  signed_at: string | null
  file_path: string | null
  created_at: string
  updated_at: string
}

export interface ContractSection { number: number; title_zh: string; title_en: string; zh: string; en: string }
export interface ContractSnapshot {
  [key: string]: unknown
  sections: ContractSection[]
  agreement_number: string
  agreement_version: number
  content_hash: string
}
export interface SignaturePoint { x: number; y: number; pressure: number }
export interface ContractSignatureRecord { agreement_id: string; agreement_version: number; agreement_content_hash: string; tenant_user_id: number; tenant_name: string; signed_at: string; property_timezone: string; consent_text_version: string; signature_hash: string; pdf_status: 'pending' | 'ready' | 'failed' }
export type ContractCategory = 'pending_effective' | 'effective' | 'expiring_soon' | 'invalid'
export interface TenantContractItem {
  agreement_id:string; agreement_number:string; agreement_version:number; agreement_content_hash:string
  order_id:string; booking_id:number; property_id:number; tenant_user_id:number; signed_at:string
  lease_start_date:string|null; lease_end_date:string|null; lease_months:number|null; property_timezone:string
  property_name:string; property_address:string; property_image_url:string|null
  payment_status:string; booking_status:string; reservation_status:string; agreement_status:string
  category:ContractCategory; category_label:string; status_labels:string[]; invalid_reason:string|null
  settlement_currency:string|null; settlement_amount_minor:number|null; payment_expires_at:string|null
  remaining_payment_seconds:number|null; remaining_contract_days:number|null; can_pay:boolean
  waiting_for_move_in:boolean; signed_pdf_available:boolean
}
export interface TenantContractDetail extends TenantContractItem { content:string; snapshot:ContractSnapshot; signature_url:string }

export const contractService = {
  listMine(): Promise<TenantContractItem[]> { return api.get('/contracts/my').then(r=>r.data.items) },
  getMine(contractId:string): Promise<TenantContractDetail> { return api.get(`/contracts/my/${contractId}`).then(r=>r.data) },
  getSignature(contractId:string): Promise<Blob> { return api.get(`/contracts/my/${contractId}/signature`, { responseType: 'blob' }).then(r=>r.data) },
  getSignedDownloadLink(contractId:string): Promise<{url?:string;expires_at?:string;code?:string;message?:string}> { return api.get(`/contracts/my/${contractId}/signed-download-link`).then(r=>r.data) },
  generate(bookingId: number): Promise<Contract> {
    return api.post(`/contracts/${bookingId}/generate`).then((r) => r.data)
  },

  get(contractId: string): Promise<Contract> {
    return api.get(`/contracts/${contractId}`).then((r) => r.data)
  },

  getByBooking(bookingId: number): Promise<Contract> {
    return api.get(`/contracts/by-booking/${bookingId}`).then((r) => r.data)
  },

  sign(contractId: string): Promise<Contract> {
    return api.post(`/contracts/${contractId}/sign`).then((r) => r.data)
  },

  confirmSignature(contractId: string, data: { agreement_version: number; agreement_content_hash: string; tenant_name: string; consent_text_version: string; idempotency_key: string; strokes: SignaturePoint[][]; name_confirmed: boolean; electronic_signature_consent: boolean }): Promise<ContractSignatureRecord> {
    return api.post(`/contracts/${contractId}/sign`, data).then((r) => r.data)
  },

  download(contractId: string): Promise<Blob> {
    return api.get(`/contracts/${contractId}/download`, { responseType: 'blob' }).then((r) => r.data)
  },
}
